src/python/gudhi/off_utils.pyx

```diff
@@ -25,21 +26,99 @@ __license__ = "MIT"
+        ValueError: If the file does not respect the OFF file format.
+    """
+    nb_vertices = -1
+
+    line = _get_next_line(file_desc)
+    # First line should be "OFF" (3d case with some variants) "4OFF" (4d case) or "nOFF" (dD case)
+    if line.lower().startswith("noff"):
+        # "nOFF" case, next line is the dimension
+        # can also contain nb_vertices, nb_faces nb_edges (can also be on the next line)
+        line = _get_next_line(file_desc)
+        digits = [int(s) for s in line.split() if s.isdigit()]
+        dim = digits[0]
+        if len(digits) > 1:
+            nb_vertices = digits[1]
+            # nb_faces =  digits[2]
+            # nb_edges =  digits[3] # not used - can be ignored
+            # nb_cells =  digits[4]
+    elif line.lower().startswith("4off"):
+        dim = 4
+    # "OFF", "COFF" and "STOFF" are 3d cases - let's stick with the C++ interface
+    elif line.lower().find("off") >= 0:
+        dim = 3
+    else:
+        raise ValueError(f"Inconsistent OFF header, got '{line.rstrip()}', should be 'OFF', '4OFF' or 'nOFF'")
+        
+    # nb_vertices can be already set by "nOFF" case, when 'dim nb_vertices nb_faces nb_edges' on the same line
+    if nb_vertices < 0:
+        # Number of points is the first number ("OFF" case) or the second one ("nOFF" case) of the second line
+        line = _get_next_line(file_desc)
+        digits = [int(s) for s in line.split() if s.isdigit()]
+        nb_vertices = digits[0]
+        # nb_faces =  digits[1]
+        # nb_edges =  digits[2] # not used - can be ignored
+        # nb_cells =  digits[3]
+    # "_get_next_line + go back to the previous line" is just a hack for comments in the most likely places
+    # TODO: remove "_get_next_line + go back to the previous line" when numpy ≥ 1.23.0 will be the standard
+    line = _get_next_line(file_desc)
+    # Here the first line without comment is read - let's go back to the beginning of this line
+    file_desc.seek(file_desc.tell() - len(line))
+    return dim, nb_vertices
+
 def read_points_from_off_file(off_file=''):
     """Read points from an `OFF file <fileformats.html#off-file-format>`_.
 
     :param off_file: An OFF file style name.
     :type off_file: string
 
     :returns:  The point set.
-    :rtype: List[List[float]]
+    :rtype: numpy.ndarray
+
+    .. warning::
+        This function is using `numpy.loadtxt <https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html>`_
+        with `comments='#'` as an argument. Empty and or comment lines between the points are only supported with numpy
+        &ge; 1.23.0.
     """
-    if off_file:
-        if os.path.isfile(off_file):
-            return read_points_from_OFF_file(off_file.encode('utf-8'))
-        else:
-            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT),
-                                    off_file)
+    # newline='' is required for Windows, otherwise end of line with '\r\n' are only detected as '\n'
```

Comment:
I don't understand the explanation. Could you explain at what point of the parsing you get what instead of what?

Reply:
On Windows, files end of lines are `\r\n`, while `\n` on Linux/Unix.
`file_desc.readline()` is always returning `my first line of file\n` on all system files when you open it with `newline=None`, cf. [open documentation](https://docs.python.org/3/library/functions.html#open) aka. universal newline mode.
As I need the exact length of the line (to go backward in the file reading), it was failing on windows because I was missing one character (the `\r`). By setting `newline=''`, I disable universal newline mode and when I read the line, I now have:
* `my first line of file\n` on Unix files
* `my first line of file\r\n` on Windows files
And I can go backward with the correct length of string.

Reply:
> As I need the exact length of the line (to go backward in the file reading),

Ah, that's the key point, you need the length of the line for seek. Could you update the comment to clarify that? It would make it clearer that we can also remove that when we drop support for old numpy.
An alternative would be to save the position (`.tell()`) before `readline` and pass that to `seek`, but that's not really better.

Reply:
Clarified on [90a5b5c](https://github.com/GUDHI/gudhi-devel/pull/1008/commits/90a5b5cd08076a9ec135534e0196ed8dcad38109)